File: src/java/4_BehavioralPatterns/3_Interpreter/InterpreterExercise.java
package com.activemesa.behavioral.interpreter.exercise;

import java.util.HashMap;
import java.util.Map;

class ExpressionProcessor
{
  public Map<Character, Integer> variables = new HashMap<>();

  enum NextOp
  {
    NOTHING, PLUS, MINUS
  }

  public static Integer tryParse(String text) {
    try {
      return Integer.parseInt(text);
    } catch (NumberFormatException e) {
      return null;
    }
  }

  public int calculate(String expression)
  {
    int current = 0;
    NextOp nextOp = NextOp.NOTHING;

    String[] parts = expression.split("(?<=[+-])");

    for (String part : parts)
    {
      String[] noOp = part.split("[+-]");
      String first = noOp[0];
      int value;

      Integer z = tryParse(first);
      if (z != null)
      {
        value = z.intValue();
      } else if (first.length() == 1
        && variables.containsKey(first.charAt(0)))
      {
        value = variables.get(first.charAt(0));
      }
      else return 0;

      switch (nextOp)
      {
        case NOTHING:
          current = value;
          break;
        case PLUS:
          current += value;
          break;
        case MINUS:
          current -= value;
          break;
      }

      if (part.endsWith("+")) nextOp = NextOp.PLUS;
      else if (part.endsWith("-")) nextOp = NextOp.MINUS;
    }
    return current;
  }
}

File: src/js/4_BehavioralPatterns/3_Interpreter/interpreter_exercise.js
class ExpressionProcessor
{
  constructor()
  {
    this.variables = {};
    this.nextOp = Object.freeze({
      nothing: 0,
      plus: 1,
      minus: 2
    });
  }

  calculate(expression)
  {
    let current = 0;
    let nextOp = this.nextOp.nothing;

    let parts = expression.split(/(?<=[+-])/);

    for (let part of parts)
    {
      let noop = part.split("+-");
      let first = noop[0];
      let value=0, z=0;

      z = parseInt(first);
      if (!isNaN(z))
        value = z;
      else if (first.length === 1
        && this.variables[first[0]] !== undefined)
      {
        value = this.variables[first[0]];
      }
      else return 0;

      switch (nextOp)
      {
        case this.nextOp.nothing:
          current = value;
          break;
        case this.nextOp.plus:
          current += value;
          break;
        case this.nextOp.minus:
          current -= value;
          break;
      }

      if (part.endsWith('+')) nextOp = this.nextOp.plus;
      else if (part.endsWith('-')) nextOp = this.nextOp.minus;
    }
    return current;
  }
}

describe('interpreter', function()
{
  it('calculate expressions with variables', function()
  {
    let ep = new ExpressionProcessor();
    ep.variables['x'] = 5;

    expect(ep.calculate('1')).toEqual(1);
    expect(ep.calculate('1+2')).toEqual(3);
    expect(ep.calculate('1+x')).toEqual(6);
    expect(ep.calculate('1+xy')).toEqual(0);
  });
});